lib/ProductOpener/CRM.pm

```diff
@@ -0,0 +1,572 @@
+
+=head1 NAME
+
+ProductOpener::CRM - manages integration with the Odoo CRM
+
+=head1 SYNOPSIS
+
+C<ProductOpener::CRM> contains functions to interact with the Odoo CRM
+
+
+=head1 DESCRIPTION
+
+For clarity:
+	- user/org refers to the pro platform side
+	- contact/company refers to the CRM side, where:
+			contact => Odoo partner is an 'individual'
+	  		company => Odoo partner is a 'company'
+
+=cut
+
+package ProductOpener::CRM;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&find_or_create_contact
+		&find_or_create_company
+		&add_contact_to_company
+		&create_opportunity
+		&add_user_to_company
+		&change_company_main_contact
+		&update_last_import_date
+		&update_last_export_date
+	);
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+
+}
+use vars @EXPORT_OK;
+
+use ProductOpener::Config2;
+use ProductOpener::Tags qw/%country_codes_reverse/;
+use ProductOpener::Users qw/retrieve_user store_user/;
+use ProductOpener::Orgs qw/retrieve_org is_user_in_org_group/;
+use XML::RPC;
+use Log::Any qw($log);
+
+my %Odoo_tags = (Producer => '10',);
+
+our @api_credentials;
+our $xmlrpc;
+
+=head2 find_or_create_contact ($user_ref)
+
+Attempts to find a contact, and if it doesn't exist, creates it
+
+=head3 Arguments
+
+=head4 $user_ref 
+
+the user to which a 'contact' in the CRM should be linked
+
+=head3 Return values
+
+the id of the contact, or undef if an error occur
+
+=cut
+
+sub find_or_create_contact($user_ref) {
+	my $contact_id = find_contact($user_ref);
+	if (defined $contact_id) {
+		return if not link_user_with_contact($user_ref, $contact_id);
+	}
+	else {
+		$contact_id = create_contact($user_ref);
+	}
+	return $contact_id;
+}
+
+=head2 link_user_with_contact ($org_ref, $contact_id = undef)
+
+Set the off_username field of a contact to the user_id  
+
+=head3 Arguments
+
+=head4 $user_ref 
+
+=head4 $contact_id 
+
+=head3 Return values
+
+1 if success, undef otherwise
+
+=cut
+
+sub link_user_with_contact($user_ref, $contact_id) {
+	my $req = Odoo('res.partner', 'write',
+		[[$contact_id], {x_off_username => $user_ref->{userid}, category_id => [[4, $Odoo_tags{Producer}]]}]);
+	$log->debug("link_user_with_contact", {user_id => $user_ref->{userid}, res => $req}) if $log->is_debug();
+	return $req;
+}
+
+=head2 find_contact ($user_ref)
+
+Finds a contact that has the user_id or same email as the given user_ref
+
+=head3 Arguments
+
+=head4 $user_ref
+
+=head3 Return values
+
+the contact id or undef
+
+=cut
+
+sub find_contact($user_ref) {
+	# find the contact with the same user_id OR email, by date of creation (oldest first)
+	my $req = Odoo(
+		'res.partner',
+		'search_read',
+		[
+			[
+				['is_company', '=', 0],
+				'|',
+				['x_off_username', '=', $user_ref->{userid}],
+				['email', '=', $user_ref->{email}]
+			]
+		],
+		{fields => ['name', 'id', 'parent_id'], order => 'create_date ASC', limit => 1}
+	);
+	my $contact = $req->[0];
```

Comment:
Why would this cause an error? If $req is undef then it should autovivify to an empty array ref so it would always be undef, or some compile flags are set that I don't know about?

Yes, if the search query has no results, it returns an empty array.